clippy_lints/src/mem_replace.rs

```diff
@@ -101,40 +101,26 @@ declare_clippy_lint! {
 impl_lint_pass!(MemReplace =>
     [MEM_REPLACE_OPTION_WITH_NONE, MEM_REPLACE_WITH_UNINIT, MEM_REPLACE_WITH_DEFAULT]);
 
-fn check_replace_option_with_none(cx: &LateContext<'_>, src: &Expr<'_>, dest: &Expr<'_>, expr_span: Span) {
-    // Check that second argument is `Option::None`
-    if is_res_lang_ctor(cx, path_res(cx, src), OptionNone) {
-        // Since this is a late pass (already type-checked),
-        // and we already know that the second argument is an
-        // `Option`, we do not need to check the first
-        // argument's type. All that's left is to get
-        // replacee's path.
-        let replaced_path = match dest.kind {
-            ExprKind::AddrOf(BorrowKind::Ref, Mutability::Mut, replaced) => {
-                if let ExprKind::Path(QPath::Resolved(None, replaced_path)) = replaced.kind {
-                    replaced_path
-                } else {
-                    return;
-                }
-            },
-            ExprKind::Path(QPath::Resolved(None, replaced_path)) => replaced_path,
-            _ => return,
-        };
-
-        let mut applicability = Applicability::MachineApplicable;
-        span_lint_and_sugg(
-            cx,
-            MEM_REPLACE_OPTION_WITH_NONE,
-            expr_span,
-            "replacing an `Option` with `None`",
-            "consider `Option::take()` instead",
-            format!(
-                "{}.take()",
-                snippet_with_applicability(cx, replaced_path.span, "", &mut applicability)
-            ),
-            applicability,
-        );
-    }
+fn check_replace_option_with_none(cx: &LateContext<'_>, dest: &Expr<'_>, expr_span: Span) {
+    // Since this is a late pass (already type-checked),
+    // and we already know that the second argument is an
+    // `Option`, we do not need to check the first
+    // argument's type. All that's left is to get
+    // the span to replacee's expr after peeling off the `&mut`
+    let sugg_expr_span = peel_ref_operators(cx, dest).span;
+    let mut applicability = Applicability::MachineApplicable;
+    span_lint_and_sugg(
+        cx,
+        MEM_REPLACE_OPTION_WITH_NONE,
+        expr_span,
+        "replacing an `Option` with `None`",
+        "consider `Option::take()` instead",
+        format!(
+            "{}.take()",
+            snippet_with_applicability(cx, sugg_expr_span, "", &mut applicability)
```

Comment:
Should be `Sugg::hir_with_context(peel_ref_operators(cx, dest), expr_span.ctxt(), "", &mut applicability).mabe_par()` to handle precedence issues and macro calls.

Reply:
Thanks~ I do like this better, but I don't know the difference between them and in what case that parenthesis are needed, could you maybe provide one example? Which I'll probably add as one of the test cases.

Reply:
`replace(&mut *x, None)` is the only practical example. If `x` is a type implementing `Deref` then the deref can't always be stripped out. This leaves you with `(*x).take()` as the required suggestion.

`x + x` could also return `&mut Option<_>`, but that's neither common nor recommended.

Reply:
Ahh, that's interesting, I completely forgot that's the case.
also, I've add a related test case